support `class` literals

alexlamsl opened the pull request.

kzc commented:
`super` is an undefined global symbol?

alexlamsl commented:
Haven't gotten that far yet. Dealing with the fall-out from usual clever design decisions &minus; specifically implicit strict mode &minus; for now.

alexlamsl commented:
So within async/generator functions, class _cannot_ use `await` or `yield` functionality, _except_ for (computed) names within the class.

How does disappointment feel again?

kzc commented:
> So within async/generator functions, class cannot use await or yield functionality, except for (computed) names within the class.

Doesn't this example contradict that?
```
$ echo 'class C{static async m(x){return 2*await x}} (async()=>{console.log(await C.m(12))})()' | node-v14.3.0 
24
```
And I thought past ufuzz failures showed that `yield` and `await` cannot be used in a computed property name. I did not try to create a test example for that.

alexlamsl commented on test/compress/classes.js:
```diff
@@ -0,0 +1,486 @@
+    expect_exact: "(async function(){return await new class{p=typeof await}})().then(function(o){console.log(o.p)});"
+    expect_stdout: "undefined"
+    node_version: ">=12"
+}
```
> Doesn't this example contradict that?

 Sorry for not being clear &minus; I meant `class` within async/generator functions, not the other way round.

kzc replied:
Now I see. That's certainly weird and unexpected. 

I wonder if it's an implementation issue or whether it's explicitly mentioned behavior in the ES specification.

alexlamsl commented:
> And I thought past ufuzz failures showed that `yield` and `await` cannot be used in a computed property name. I did not try to create a test example for that.

I think that discussion was for destructuring objects &minus; plain object literal works:
```sh
$ echo '(async function() { return {[await 4]: 2} })().then(console.log)' | node
{ '4': 2 }
```

kzc commented:
> I think that discussion was for destructuring objects − plain object literal works:

I can't remember all the ES6+ features and exceptions.  
:-/

alexlamsl commented:
> I can't remember all the ES6+ features and exceptions.

... the secret to a happy life 🤣

kzc commented:
Hmm... AST_Super is not derived from AST_Class as we did in uglify-es? We didn't need to even mention AST_Super in lib/compress.js as result. Is there a scenario where the two classes behave differently in the Compressor?

alexlamsl commented:
Merely because of this quick observation:
```sh
$ echo 'this' | node

$ echo 'this' | uglify-js -c


$ echo 'super' | node
[stdin]:1
super
^^^^^

SyntaxError: 'super' keyword unexpected here
    at [stdin]-wrapper:6:22

$ echo 'super' | uglify-js -c
super;
```
But just like `import.meta`, turns out this thing isn't a normal `AST_Symbol` either &minus; that's why I've basically block in from most optimisations while I sort the other `class`-related parts out.

kzc commented:
`super` parse issues and where AST_Super can be instantiated via `new` are already a special case - is it in a method, or object literal, etc. But once instantiated it behaves like `this` in the great majority of cases and is subject to the same Compressor optimizations and exclusion logic. Alternatively, AST_Super and AST_This could have a common base class - AST_ObjectIdentity or something - to save CPU in fewer `instanceof` checks. Every new type introduced makes the Compressor ever so slightly slower due to the numerous if-type-chains.

alexlamsl commented:
Thanks for the analysis &minus; `AST_ObjectIdentity` introduced in https://github.com/mishoo/UglifyJS/pull/4658/commits/f3e91cc81367f5dd4785ca035237d977529e8e81

alexlamsl commented on test/compress/classes.js:
```diff
@@ -0,0 +1,574 @@
+        console.log("A", A.p, A.q);
+        console.log("B", B.p, B.q);
+        new B();
+    }
```
If you replace those getters with fields, the result will surprise ye~

kzc replied:
I tried but didn't see anything unusual. Do you mean that static properties are inherited from the base?

alexlamsl replied:
Original output:
```
A Foo 42
B Bar Foo
a.p undefined bar
a.q undefined foo
b.p foo bar
b.q null foo
```
<details>
<summary>Replace getters with fields:</summary>

```js
"use strict";
class A {
    static p = "Foo";
    static q = super.p || 42;
    constructor() {
        console.log("a.p", super.p, this.p);
        console.log("a.q", super.q, this.q);
    }
    p = "foo";
    q = super.p || null;
}
class B extends A {
    static p = "Bar";
    static q = super.p;
    constructor() {
        super();
        console.log("b.p", super.p, this.p);
        console.log("b.q", super.q, this.q);
    }
    p = "bar";
    q = super.p;
}
console.log("A", A.p, A.q);
console.log("B", B.p, B.q);
new B();
```

</details>

Replaced output:
```
A Foo 42
B Bar Foo
a.p undefined foo
a.q undefined null
b.p undefined bar
b.q undefined undefined
```

kzc replied:
Ah those class fields. I don't think they're in the ES standard yet... https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields, although most browsers and tools (rollup, esbuild, babel) implement some subset of the proposed features.

I think the way class instance fields work is prepending the initialisations to the constructor.  I don't know how the class static ones are supposed to be initialized. Different implementations do not agree on the use of `super` in field initialisations.

alexlamsl replied:
In the ideal world you'd hope they make it work the same way as in object literals &minus; not that the latter is working intuitively IMO

alexlamsl commented:
*TIL*
```js
var o = {};
Object.setPrototypeOf({
    f() { super.p = 42; }
}, o).f();
console.log("p" in o, o.p);
```
▶️ `false undefined`
```js
var o = {
    set p(v) { console.log(v); },
};
Object.setPrototypeOf({
    f() { super.p = 42; },
}, o).f();
```
▶️ `42`

kzc commented:
Your *TIL*s hurt my brain. 

fwiw, it's consistent...

```
$ cat til.js
var base = {};
var obj = Object.setPrototypeOf({
    f() { super.p = 42; }
}, base);
obj.f();
console.log("base:", "p" in base, base.p);
console.log("obj:", "p" in obj, obj.p);
```
```
$ cat til.js | node
base: false undefined
obj: true 42
```
The changed property is not set in the "base" prototype object, but the newly merged object returned from `Object.setPrototypeOf`.

alexlamsl commented:
> The changed property is not set in the "base" prototype object, but the newly merged object returned from `Object.setPrototypeOf`.

... except when your base object has a setter for that property? 😏

alexlamsl commented:
I see what you mean:
```js
var o = Object.setPrototypeOf({
    f() {
        super.p = 42;
    },
    g() {
        return super.p;
    },
},{});
o.f();
console.log(o.p, o.g());
```
▶️ `42 undefined`

But in that case, what's the difference between `super.p` and `this.p`? 🤔

alexlamsl commented:
> Your TILs hurt my brain.

Hey I didn't write those ~job interview questions~ language specifications... 😏

kzc commented:
> ... except when your base object has a setter for that property?

It's consistent. The setter is just a function. The "this" the setter refers to is dependent on the context in which it is called. In your example you just had a print. The property was not set anywhere. Which object the property is set is the issue.

```
$ cat til2.js 
var base = {
    set dump(v) {
        console.log("dump:", "p" in this, this.p);
    }
};
var obj = Object.setPrototypeOf({
    f() { super.p = 42; }
}, base);
obj.f();
console.log("base:", "p" in base, base.p);
console.log("obj:", "p" in obj, obj.p);
obj.dump = 123;
```
```
$ cat til2.js | node
base: false undefined
obj: true 42
dump: true 42
```
The setter is indeed called, but its "this" is "obj", not "base".

alexlamsl commented:
Then the question remains &minus; what's the difference between `super.p = 42` and `this.p = 42`?

Because if the answer is that they are identical, then I'd like to... poke... whoever write that spec...

kzc commented:
> But in that case, what's the difference between super.p and this.p?

```
$ cat til3.js 
var base = {
    p: 123,
};
var obj = Object.setPrototypeOf({
    p: 456,
    print() {
        console.log("super.p:", super.p);
        console.log("this.p:", this.p);
    }
}, base);
obj.print();
```
```
$ cat til3.js | node
super.p: 123
this.p: 456
```
Reading still respects the prototype chain. However, when properties are set they go into the instance - the prototype chain objects are not altered.

alexlamsl commented:
> Reading still respects the prototype chain.

Now if only class literals work the same way... 😏

alexlamsl commented:
```js
var b = { p: 1 };
var o = Object.setPrototypeOf({
    print() {
        console.log(super.p++, this.p);
    },
    p: "FAIL",
}, b);
o.print();
o.print();
b.p = 10;
o.print();
```
▶️
```
1 2
1 2
10 11
```
So your claim of read from base but assign to instance is working above (not to say it's intuitive or consistent, but hey...) &minus; however if the base object has a setter instead:
```js
var a = 1;
var b = {
    get p() {
        return a;
    },
    set p(v) {
        console.log('SET', v);
        a = v;
    },
};
var o = Object.setPrototypeOf({
    print() {
        console.log(super.p++, this.p);
    },
    p: 42,
}, b);
o.print();
o.print();
```
▶️
```
SET 2
1 42
SET 3
2 42
```
So the assignment does get passed to the base object here, does it not? 🤔

kzc commented:
> Now if only class literals work the same way... 

Point taken with instance variables, but with statics they do...

```
$ cat class1.js 
class Base {
    static prop = "Base";
    static meth() {
        return "Base";
    }
}
class Derived extends Base {
    static prop = "Derived";
    static meth() {
        return "Derived";
    }
    static print() {
        console.log("super.meth() =", super.meth());
        console.log("this.meth() =", this.meth());
        console.log("super.prop =", super.prop);
        console.log("this.prop =", this.prop);
    }
}
Derived.print();
```
```
$ cat class1.js | node
super.meth() = Base
this.meth() = Derived
super.prop = Base
this.prop = Derived
```

The ES5 `super` example via `Object.setPrototypeOf` was essentially dealing with static objects rather than instance objects.

kzc commented:
> So the assignment does get passed to the base object here, does it not? 🤔

Your example does not demonstrate that the base object is assigned to. It only demonstrates the setter *function* was called and sets a global variable. Have the setter set a different property on `this`.

alexlamsl commented:
> The setter is just a function. The "this" the setter refers to is dependent on the context in which it is called.

By that logic the (intrinsic) getter should reference the same `this` as well, which means:
```js
var base = {
    get p() {
        return this.p;
    },
    set p(v) {
        this.p = v;
    },
};
```
obviously `super.p` won't work at all if that's the case 😓

... which implies getter & setter for `super` have different context, i.e. base & instance respectively. Not my definition of being consistent here.

kzc commented:
> obviously super.p won't work at all if that's the case

I think you'll get infinite recursion with that example. You probably want to set `this._p = v`.

It's consistent. Review the examples again. Prototype inheritance would not work at all if `this.prop` assignments altered the prototype chain. The changes are set in the instance.

alexlamsl commented:
> It's consistent. Review the examples again. 

Shouldn't this:
```js
var a = 1;
var b = {
    get p() {
        return a;
    },
    set p(v) {
        console.log('SET', v);
        a = v;
    },
};
```
be functionally equivalent to this?
```js
var b = { p: 1 };
```

kzc commented:
It's equivalent in that very specific case, but not in general. Setters and getters are just functions with a `this` context. If you're not using the `this` context to set properties you will not see that the object context for `this` changes to the derived instance. As for the getter, it starts with the derived instance and then goes up the prototype chain until it finds that property name.

alexlamsl commented:
Here's an example that uses `this`:
```js
var base = {
    get prop() {
        return this.store;
    },
    set prop(value) {
        this.store = value;
    },
};
base.prop = 1;
var instance = Object.setPrototypeOf({
    print() {
        console.log(super.prop++, this.prop);
    },
    prop: 42,
}, base);
instance.print();
instance.print();
```
▶️
```
1 42
2 42
```
Here's the same code with getter/setter replaced by ordinary property:
```js
var base = {
    prop: 1,
};
var instance = Object.setPrototypeOf({
    print() {
        console.log(super.prop++, this.prop);
    },
    prop: 42,
}, base);
instance.print();
instance.print();
```
▶️
```
1 2
1 2
```

kzc commented:
In the store example you probably wanted to write `base.store = 1;` instead of `base.prop = 1;`, but no matter.

Notice the base object remains unchanged with store = 1 in the setter example, and store = 3 in instance:

```
$ cat setter.js 
var base = {
    get prop() {
        return this.store;
    },
    set prop(value) {
        this.store = value;
    },
};
base.store = 1;
var instance = Object.setPrototypeOf({
    print() {
        console.log(super.prop++, this.prop);
    },
    prop: 42,
}, base);
instance.print();
instance.print();
console.log("base:", base);
console.log("instance:", instance);
```
```
$ cat setter.js | node
1 42
2 42
base: { prop: [Getter/Setter], store: 1 }
instance: { print: [Function: print], prop: 42, store: 3 }
```
and prop remains 1 in base in the ordinary example...
```
$ cat ordinary.js 
var base = {
    prop: 1,
};
var instance = Object.setPrototypeOf({
    print() {
        console.log(super.prop++, this.prop);
    },
    prop: 42,
}, base);
instance.print();
instance.print();
console.log("base:", base);
console.log("instance:", instance);
```
```
$ cat ordinary.js | node
1 2
1 2
base: { prop: 1 }
instance: { print: [Function: print], prop: 2 }
```

kzc commented:
@alexlamsl I don't pretend to really understand ECMAScript, I'm just figuring it out as I go along. I could easily be wrong. Someone familiar with the specification would probably get a good laugh out of my superficial (mis)understanding of it.

alexlamsl commented:
Oh don't worry about that &minus; I'm not sure the original authors of those language features know any better than you & me here 😏

What I'm venting here is merely how they created new corner cases that makes little common sense. Just like pure mathematics, one can write logically consistent spec that are extremely counter-intuitive. However one should then question whether there is value in using such a creation.

kzc commented:
The following example shows that `super` does behave like `this` for property assignments, and how it behaves for super method invocations:

```
$ cat method.js 
class Base {
    constructor(x) {
        this.x = x;
    }
    method() {
        console.log("Base#method:", this);
    }
}
class Derived extends Base {
    constructor(x, y) {
        super(x);
        this.foo = 123;
        super.foo = 456; // will overwrite this.foo assignment above
        this.bar = 789;
        this.y = y;
        this.method();
        super.method();
    }
    method() {
        console.log("Derived#method:", this);
    }
}
new Derived("first", "second");
```
```
$ cat method.js | node-v14.3.0 
Derived#method: Derived { x: 'first', foo: 456, bar: 789, y: 'second' }
Base#method: Derived { x: 'first', foo: 456, bar: 789, y: 'second' }
```

alexlamsl commented:
Will make another release before merging this, but I think after this PR we can declare `uglify-js` as a proper replacement for `uglify-es`? 🤔

kzc commented:
Congratulations. With this PR I think you've exceeded the ES6+ support in uglify-es.

I suppose you could test against uglify-es' test suite and ignore the `expect:`, `expect_exact:` and `expect_warnings:` blocks and only verify against `expect_stdout:`. Locally my test runner checks against `+process.env.IGNORE_EXPECT` to do just that. 

I'm sure some issues will also be found if large ES6+ code bases are minified in-situ and running their test suites. Minifying their bundles isn't sufficient - it won't exercise import/export AST rename+compress+mangle.

kzc commented:
Found testing against the last ES version of Rollup before the Typescript changeover...
```
$ curl -s https://raw.githubusercontent.com/rollup/rollup/3d80c06f895eab41e648ee99786fa68c72458b80/src/utils/identifierHelpers.js | bin/uglifyjs
Parse error at STDIN:11,31
const startsWithDigit = str => /\d/.test( str[0] );
                               ^
ERROR: Unexpected token: operator «/»
```
Standalone test case:
```
$ echo 'const d = s => /\d/.test(s[0]); console.log(d("5C"), d("A2"));' | node
true false
```
```
$ echo 'const d = s => /\d/.test(s[0]); console.log(d("5C"), d("A2"));' | bin/uglifyjs
Parse error at STDIN:1,15
const d = s => /\d/.test(s[0]); console.log(d("5C"), d("A2"));
               ^
ERROR: Unexpected token: operator «/»
```

kzc commented:
Some ES code bases rely on the implicit ES6 top level `"use strict"` behavior when no directives are specified. I'd recommend supporting a `--module` option for CLI, compress and mangle as previously mentioned in https://github.com/mishoo/UglifyJS/issues/4661#issuecomment-781666319. Otherwise uglify-js could run into scoping issues.

kzc commented:
Another bug found with the rollup source tree...

Given:
```
$ curl -s https://raw.githubusercontent.com/rollup/rollup/3d80c06f895eab41e648ee99786fa68c72458b80/src/ast/scopes/BlockScope.js
import Scope from './Scope';

export default class BlockScope extends Scope {
	addDeclaration ( identifier, options = {} ) {
		if ( options.isHoisted ) {
			return this.parent.addDeclaration( identifier, options );
		} else {
			return super.addDeclaration( identifier, options );
		}
	}
}
```
`uglify-js` with no compress, no mangle is fine:
```
$ curl -s https://raw.githubusercontent.com/rollup/rollup/3d80c06f895eab41e648ee99786fa68c72458b80/src/ast/scopes/BlockScope.js | bin/uglifyjs | acorn --module --ecma2020 --silent && echo SUCCESS || echo FAIL
SUCCESS
```
Output with `--toplevel -mc`:
```
$ curl -s https://raw.githubusercontent.com/rollup/rollup/3d80c06f895eab41e648ee99786fa68c72458b80/src/ast/scopes/BlockScope.js | bin/uglifyjs --toplevel -mc
import"./Scope";export default class extends(void 0){addDeclaration(e,t={}){return(t.isHoisted?this.parent:super).addDeclaration(e,t)}};
```
`acorn` has an issue with the optimization on `super`:
```
$ curl -s https://raw.githubusercontent.com/rollup/rollup/3d80c06f895eab41e648ee99786fa68c72458b80/src/ast/scopes/BlockScope.js | bin/uglifyjs --toplevel -mc | acorn --module --ecma2020
Unexpected token (1:112)
```
When uglify-js output is verified with a different ES parser, `esbuild`, it reports the same issue:
```
$ curl -s https://raw.githubusercontent.com/rollup/rollup/3d80c06f895eab41e648ee99786fa68c72458b80/src/ast/scopes/BlockScope.js | bin/uglifyjs --toplevel -mc | esbuild
 > <stdin>: error: Unexpected "super"
    1 │ ...ass extends(void 0){addDeclaration(e,t={}){return(t.isHoisted?this.parent:super).addDeclaration(e,t)}};
      ╵                                                                              ~~~~~

1 error
```
So apparently `super` has to be excluded from some compress optimizations.

alexlamsl commented:
> `const startsWithDigit = str => /\d/.test( str[0] );`

I can't reproduce this issue &minus; are you using the latest release instead of `master` or this PR?

I think this is fixed by https://github.com/mishoo/UglifyJS/commit/5f60c1b830fec648974b3ea7cbd6f53a6bfd7c75, which is yet to be released (but is included in this PR's branch)

kzc commented:
> I can't reproduce this issue − are you using the latest release instead of master or this PR?

I applied this PR patch to local uglify-js master.  If your branch works, that's fine.

alexlamsl commented:
> `return (t.isHoisted ? this.parent : super).addDeclaration(e, t);`

Oh oops &minus; that's what I was worried about in https://github.com/mishoo/UglifyJS/pull/4658#issuecomment-782302004 &minus; time to patch up...

kzc commented:
> Oh oops − that's what I was worried about in #4658 (comment) − time to patch up...

I think the strategy of having a common base class of AST_This and AST_super is sound.  There was always going to be exceptions. If `conditionals` is disabled it works in this particular case.

kzc commented:
Another one...
```
$ echo 'import("./foo"); import("./foo/" + "bar");' | bin/uglifyjs
Parse error at STDIN:1,6
import("./foo"); import("./foo/" + "bar");
      ^
ERROR: Unexpected token: punc «(», expected: punc «{»
```

alexlamsl commented:
> `import("./foo"); import("./foo/" + "bar");`

I've dealt with the `var a = import("./foo");` case as you said above, just forgot to make a fallback in the parser for when statement starts with the `import` keyword...

kzc commented:
If each of the 136 individual source files within `rollup/src` at rollup commit `3d80c06f895eab41e648ee99786fa68c72458b80` is minified in-situ with uglify-js with compress and mangle disabled and `npm test` is run it results in 5 mocha test failures. 

If each source file is minified in-situ with `uglify-js -mc conditionals=0` it results in `1523 passing (11s), 3 pending, 18 failing` mocha tests.

If `-c toplevel` is used then `npm test` fails to run at all due to this issue:
```
$ echo "import {bar} from './whatever'; const foo=bar();" | uglify-js -c toplevel
import{}from"./whatever";(void 0)();
```

kzc commented:
@alexlamsl Please takeover testing against the rollup source tree with the following patch and bash script:
```patch
diff --git a/rollup.config.js b/rollup.config.js
index b685b81..423c531 100644
--- a/rollup.config.js
+++ b/rollup.config.js
@@ -1,5 +1,4 @@
 import { readFileSync } from 'fs';
-import buble from 'rollup-plugin-buble';
 import resolve from 'rollup-plugin-node-resolve';
 import commonjs from 'rollup-plugin-commonjs';
 import json from 'rollup-plugin-json';
@@ -25,12 +24,6 @@ export default [
 		input: 'src/node-entry.js',
 		plugins: [
 			json(),
-			buble({
-				include: ['src/**', 'node_modules/acorn/**'],
-				target: {
-					node: '4'
-				}
-			}),
 			resolve(),
 			commonjs()
 		],
@@ -48,12 +41,6 @@ export default [
 		input: 'src/browser-entry.js',
 		plugins: [
 			json(),
-			buble({
-				include: ['src/**', 'node_modules/acorn/**'],
-				target: {
-					node: '4'
-				}
-			}),
 			resolve(),
 			commonjs(),
 			{
@@ -80,7 +67,6 @@ export default [
 		plugins: [
 			string({ include: '**/*.md' }),
 			json(),
-			buble({ target: { node: 4 } }),
 			commonjs({
 				include: 'node_modules/**'
 			}),
diff --git a/src/utils/identifierHelpers.js b/src/utils/identifierHelpers.js
index 654f017..3a9141e 100644
--- a/src/utils/identifierHelpers.js
+++ b/src/utils/identifierHelpers.js
@@ -8,7 +8,7 @@ reservedWords.concat( builtins ).forEach( word => blacklisted[ word ] = true );
 
 const illegalCharacters = /[^$_a-zA-Z0-9]/g;
 
-const startsWithDigit = str => /\d/.test( str[0] );
+const startsWithDigit = str => (/\d/.test( str[0] ));
 
 export function isLegal ( str ) {
 	if ( startsWithDigit(str) || blacklisted[ str ] ) {
diff --git a/test/mocha.opts b/test/mocha.opts
index b165c53..4e8a550 100644
--- a/test/mocha.opts
+++ b/test/mocha.opts
@@ -1,2 +1 @@
---require buble/register
 test/test.js
```
```
$ cat test-rollup.sh 
#!/bin/sh
rm -rf rollup
git clone https://github.com/rollup/rollup
cd rollup
git checkout 3d80c06f895eab41e648ee99786fa68c72458b80
rm -rf node_modules
npm ci
patch -p1 < ../rollup.patch
# UGLIFY_OPTIONS=" --toplevel -mc conditionals=0,passes=9,pure_getters,unsafe "
UGLIFY_OPTIONS=" -mc conditionals=0,passes=9,pure_getters,unsafe "
for i in `find src -name '*.js'`
do
    echo "$i"
    cp -f "$i" "$i.bak"
    uglify-js "$i.bak" $UGLIFY_OPTIONS > "$i" && rm -f "$i.bak"
done
npm test \
    && echo "SUCCESS with UGLIFY_OPTIONS='$UGLIFY_OPTIONS'" \
    || echo  "FAILED with UGLIFY_OPTIONS='$UGLIFY_OPTIONS'"
```

Just run `bash test-rollup.sh`. 
This produced 18 test failures for my local version of `uglify-js`.
I did not investigate any further.
`uglify-es` passed all tests as I recall.
Change `UGLIFY_OPTIONS` and retest as necessary.

alexlamsl commented:
Thanks a lot &minus; will do 👌

kzc commented:
It should be possible to use the same technique to test other major ES projects like `acorn` or even the latest `rollup` tree. However you'd have to add a line in the bash for-loop to convert Typescript to ES using a tool of your choice like esbuild or babel.

kzc commented:
FWIW, when I tested that script on rollup years ago with uglify-es, I also found an acorn bug and several rollup bugs.

kzc commented:
Please note: that old version of rollup did not exercise `async` functionality like the latest version of the rollup source tree does. Once converted from Typescript it should be a very useful test case for uglify-js.

alexlamsl commented:
patch:
```diff
diff --git a/package.json b/package.json
index 4838ac361..0c50d742c 100644
--- a/package.json
+++ b/package.json
@@ -18,7 +18,7 @@
     "test-coverage": "rm -rf coverage/* && istanbul cover --report json node_modules/.bin/_mocha -- -u exports -R spec test/test.js",
     "posttest-coverage": "remap-istanbul -i coverage/coverage-final.json -o coverage/coverage-remapped.json -b dist && remap-istanbul -i coverage/coverage-final.json -o coverage/coverage-remapped.lcov -t lcovonly -b dist && remap-istanbul -i coverage/coverage-final.json -o coverage/coverage-remapped -t html -b dist",
     "ci": "npm run test-coverage && codecov < coverage/coverage-remapped.lcov",
-    "build": "git rev-parse HEAD > .commithash && rollup -c && chmod a+x bin/rollup",
+    "build": "git rev-parse HEAD > .commithash && rollup -c",
     "watch": "rollup -cw",
     "prepublishOnly": "npm run lint && npm run test:only && npm run test:leak",
     "prepare": "npm run build",
diff --git a/rollup.config.js b/rollup.config.js
index b685b8144..423c5312e 100644
--- a/rollup.config.js
+++ b/rollup.config.js
@@ -1,5 +1,4 @@
 import { readFileSync } from 'fs';
-import buble from 'rollup-plugin-buble';
 import resolve from 'rollup-plugin-node-resolve';
 import commonjs from 'rollup-plugin-commonjs';
 import json from 'rollup-plugin-json';
@@ -25,12 +24,6 @@ export default [
                input: 'src/node-entry.js',
                plugins: [
                        json(),
-                       buble({
-                               include: ['src/**', 'node_modules/acorn/**'],
-                               target: {
-                                       node: '4'
-                               }
-                       }),
                        resolve(),
                        commonjs()
                ],
@@ -48,12 +41,6 @@ export default [
                input: 'src/browser-entry.js',
                plugins: [
                        json(),
-                       buble({
-                               include: ['src/**', 'node_modules/acorn/**'],
-                               target: {
-                                       node: '4'
-                               }
-                       }),
                        resolve(),
                        commonjs(),
                        {
@@ -80,7 +67,6 @@ export default [
                plugins: [
                        string({ include: '**/*.md' }),
                        json(),
-                       buble({ target: { node: 4 } }),
                        commonjs({
                                include: 'node_modules/**'
                        }),
diff --git a/test/mocha.opts b/test/mocha.opts
index b165c5357..4e8a550a4 100644
--- a/test/mocha.opts
+++ b/test/mocha.opts
@@ -1,2 +1 @@
---require buble/register
 test/test.js
```
running `npm ci` then `npm test` on Node.js v15 (Windows x64) gives:
```
  1528 passing (23s)
  3 pending
  3 failing
```

kzc commented:
What results do you get with:
```
UGLIFY_OPTIONS=" --toplevel -mc passes=9,pure_getters,unsafe "
```

alexlamsl commented:
Will get there in a moment &minus; even with the default options I'm getting:
```
  1425 passing (17s)
  3 pending
  106 failing
```

alexlamsl commented:
With your options it failed like this... 😅 
```
import{}from"./object.js";const e="break case class catch const continue debugger default delete do else export extends finally for function if import in instanceof let new return super switch this throw try typeof var void while with yield enum await implements package protected static interface private public".split(" ");var r="Infinity NaN undefined null true false eval uneval isFinite isNaN parseFloat parseInt decodeURI decodeURIComponent encodeURI encodeURIComponent escape unescape Object Function Boolean Symbol Error EvalError InternalError RangeError ReferenceError SyntaxError TypeError URIError Number Math Date String RegExp Array Int8Array Uint8Array Uint8ClampedArray Int16Array Uint16Array Int32Array Uint32Array Float32Array Float64Array Map Set WeakMap WeakSet SIMD ArrayBuffer DataView JSON Promise Generator GeneratorFunction Reflect Proxy Intl".split(" ");const t=(void 0)();e.concat(r).forEach(e=>t[e]=!0);const a=/[^$_a-zA-Z0-9]/g;var n=e=>/\d/.test(e[0]);function o(e){return!n(e)&&!t[e]&&!a.test(e)}function i(e){return e=e.replace(/-(\w)/g,(e,r)=>r.toUpperCase()).replace(a,"_"),n(e)||t[e]?`_${e}`:e}export{e as reservedWords,o as isLegal,i as makeLegal};
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               ^
TypeError: (void 0) is not a function
    at Object.<anonymous>
```

kzc commented:
> 3 failing

Ah, okay - you mean with no minify.  When I last tested uglify-es with this I used an older version of node.

On Mac I have 5 `npm test` failures with node-v14.3.0 and 3 failures with node v10.4.1.

With node v8.0.0:

```
  1541 passing (20s)
  3 pending
```

kzc commented:
> With your options it failed like this...

It's the uglify-js bug mentioned above:

https://github.com/mishoo/UglifyJS/pull/4658#issuecomment-782737002

alexlamsl commented:
This isn't the `super` issue (which has been fixed), it's `AST_SymbolImport` getting initialised with improperly, and `reduce_vars` ended up replacing its reference with `void 0` 😅

kzc commented:
with refreshed #4658 - see `extends (void 0)`:
```
$ curl -s https://raw.githubusercontent.com/rollup/rollup/3d80c06f895eab41e648ee99786fa68c72458b80/src/ast/scopes/BlockScope.js | bin/uglifyjs --toplevel -bc
import "./Scope";

export default class extends (void 0) {
    addDeclaration(identifier, options = {}) {
        return (options.isHoisted ? this.parent.addDeclaration : super.addDeclaration)(identifier, options);
    }
};
```

alexlamsl commented:
After patching up that `import` bug:
```
  1425 passing (17s)
  3 pending
  106 failing
```

alexlamsl commented:
Yeah it's got to be Node.js v8 &minus; even v10 will produce test errors (without `uglify-js`, that is)

alexlamsl commented:
Okay fixing one more bug brings it down to:
```
  1517 passing (22s)
  3 pending
  14 failing
```
Now here's the funny part &minus; the code that seems to be failing does:
```js
export function addTask(...) {
    ... new FileWatcher(...) ...
}
...
export default class FileWatcher {
    ...
}
```
I thought `export default` can only hold an expression rather than a statement?! 🤨

kzc commented:
> I thought export default can only hold an expression rather than a statement?! 🤨

No, a single var/let/const statement and class and function declarations can exported as well.  When in doubt use `acorn --module --ecma2020` to test.

alexlamsl commented:
Confirmed &minus; if I replace that with:
```js
class FileWatcher {
    ...
}
export default FileWatcher;
```
all tests passed (well, 3 pending) with `-mc`

alexlamsl commented:
> No, a single var/let/const statement and class and function declarations can exported as well. 

I'm talking about `export default`, not `export`:
```js
// test.mjs
export var a = 1;
export default var b = 2;
```
```sh
$ node test.mjs
test.mjs:2
export default var b = 2;
               ^^^

SyntaxError: Unexpected token 'var'
```

alexlamsl commented:
Apparently any unnamed class/function are expressions, but if it gets a name it's a `DefClass`/`Defun` &minus; IE8 must be so proud right now 😏

alexlamsl commented:
Which means we now get this 💎:
```js
//test.mjs
export default class{}console.log("PASS");
```
```sh
$ node test.mjs
PASS
```

alexlamsl commented:
```js
// fail.mjs
export default 42 console.log("PASS");
```
```sh
$ node fail.js
fail.mjs:1
export default 42 console.log("PASS");
                  ^^^^^^^

SyntaxError: Unexpected identifier
```

```js
// call.mjs
export default async function() {
    console.log("FAIL");
}(console.log("PASS"));
```
```sh
$ node call.mjs
PASS
```

kzc commented:
I ran into this bug. It's a variation of what was mentioned above.

```
$ echo 'export default class C{};C.prototype.p=1;' | uglify-js --toplevel -bmc
export default class {};

C.prototype.p = 1;
```
```
$ echo 'export default function F(){};F.prototype.p=1;' | uglify-js --toplevel -bmc
export default function() {};

F.prototype.p = 1;
```

The necessary class and function names were dropped.

kzc commented:
Okay, the latest version of the rollup source tree can be tested against uglify-js with the help of esbuild ts to js translation...

`rollup-convert-ts.patch:`
```patch
diff --git a/package.json b/package.json
index a9df600..4dcbc5c 100644
--- a/package.json
+++ b/package.json
@@ -35 +35 @@
-    "test:all": "npm run test:only && npm run test:browser && npm run test:typescript && npm run test:leak && npm run test:package",
+    "test:all": "npm run test:only && npm run test:browser && npm run test:leak && npm run test:package",
```
`test-rollup-v2.39.0.sh`:
```
#!/bin/sh

rm -rf rollup
git clone --depth 1 --branch v2.39.0 https://github.com/rollup/rollup
rm -rf rollup/.git/hooks
cd rollup
rm -rf node_modules
npm ci
npm i esbuild
rm -rf dist
patch -p1 < ../rollup-convert-ts.patch
# UGLIFY_OPTIONS="--toplevel -mc passes=9,pure_getters,unsafe"
# UGLIFY_OPTIONS="--toplevel -mc"
# UGLIFY_OPTIONS="-mc "
UGLIFY_OPTIONS=""
for i in `find cli src -name '*.ts' | grep -v '\.d\.ts'`
do
    echo "$i"
    cp -f "$i" "$i.bak"
    node_modules/.bin/esbuild --loader=ts --target=node14 < "$i.bak" \
        | uglify-js $UGLIFY_OPTIONS > "$i" && rm -f "$i.bak"
done
npm test \
    && echo "SUCCESS with UGLIFY_OPTIONS='$UGLIFY_OPTIONS'" \
    || echo  "FAILED with UGLIFY_OPTIONS='$UGLIFY_OPTIONS'"
```
It passes all tests using node v14.3.0 with uglify-js whitespace minification only.

alexlamsl commented:
Right, got https://github.com/mishoo/UglifyJS/pull/4658#issuecomment-782763045 to pass all tests under every `test/ufuzz/options.json` settings (on Node.js v8)

kzc commented:
`acorn` does not use Typescript as I thought - just plain old ES with js suffixes. In theory it should be trivial to perform in-situ minify testing, but it has weird map errors with just whitespace elimination possibly pointing to the wrong place. At one time I was able to successfully test uglify-es against its test suite but its build has become more complex since that time. I'll leave that to you - I give up.

The more projects uglify-js can be tested against, the more robust it will be. Every project exercises different parts of the ES spec.

alexlamsl commented:
Currently trying out the `esbuild` recipe &minus; without any `esbuild` or `uglify-js` that latest version of `rollup` gives (on Node.js v15):
```
  4091 passing (4m)
  2 pending
  7 failing
```

alexlamsl commented:
... same result on Node.js v14 &minus; may be it's a Windows thing 🤔

kzc commented:
> ... same result on Node.js v14 − may be it's a Windows thing 🤔

Zero failures on Mac with no uglify-js CLI options with Node v14.3.0.

But you're saying that Rollup won't "npm test" cleanly on Windows after a `git clone && npm ci`?  I do recall that a number of tests are skipped on Windows - not sure how they are reported in the summary.

Are you able to use Windows Subsystem for Linux (odd name - seems backwards) to run the unix tests?

alexlamsl commented:
> But you're saying that Rollup won't "npm test" cleanly on Windows after a git clone && npm ci?

Indeed &minus; looking at the assertion errors, at least a few of them is due to the `\` vs `/` in path names

alexlamsl commented:
> Are you able to use Windows Subsystem for Linux (odd name - seems backwards) to run the unix tests?

I've been using VirtualBox for so long now, haven't had aneed for WSL (clearly not a name came up by their marketing department 👻) yet.

kzc commented:
How many failures are you seeing with latest rollup on VirtualBox/Linux when its sources are filtered through uglify-js?

alexlamsl commented:
No failures on Linux &minus; as suspected, it's the path separator differences, plus a couple of tests for symlinks (don't have those on NTFS).

kzc commented:
This PR looks good to go. Do you think it needs more testing and fuzzing?

alexlamsl commented:
Thanks for the vote of confidence 👻

Yes, I'm planning to weed out more false positives in the Fuzzer before merging, but the main code should be in its final form.

kzc commented:
@alexlamsl I haven't run `npm test` in a while, so I tried it [v3.13.0](bd737200615551e49d929fd6a55c3c71362ec507) using node v14.3.0:

```
--- classes.js
    Running test [constructor_1]
    Running test [constructor_2]
    Running test [constructor_3]
    Running test [constructor_4]
    Running test [fields]
    Running test [methods]
    Running test [private_methods]
!!! Invalid input or expected stdout
---INPUT---
{
    new class A {
        static *#f() {
            yield A.#p * 3;
        }
        async #g() {
            for (var a of A.#f()) return a * await 2;
        }
        static get #p() {
            return 7;
        }
        get q() {
            return this.#g();
        }
    }().q.then(console.log);
}
---EXPECTED STDOUT---
42

---ACTUAL ERROR---
SyntaxError: Unexpected token '('
```

I didn't realize the private member support was also added - very cool.

A small tweak is needed to pass tests:

```patch
--- a/test/compress/classes.js
+++ b/test/compress/classes.js
@@ -125,7 +125,7 @@ private_methods: {
     }
     expect_exact: "(new class A{static*#f(){yield 3*A.#p}async #g(){for(var a of A.#f())return a*await 2}static get #p(){return 7}get q(){return this.#g()}}).q.then(console.log);"
     expect_stdout: "42"
-    node_version: ">=14"
+    node_version: ">=15"
 }
 
 await: {
```

Consider adding node 14 to the test matrix:

https://github.com/mishoo/UglifyJS/blob/bd737200615551e49d929fd6a55c3c71362ec507/.github/workflows/ci.yml#L10-L12

BTW, congrats on the v3.13.0 release - it's a huge jump in ES support.

alexlamsl commented:
I guess they've added support half-way through v14: 🤔 
```
$ node -v
v14.15.5

$ node test/compress classes.js
--- classes.js
    Running test [constructor_1]
    Running test [constructor_2]
    Running test [constructor_3]
    Running test [constructor_4]
    Running test [fields]
    Running test [methods]
    Running test [private_methods]
    Running test [await]
    Running test [yield]
    Running test [conditional_parentheses]
    Running test [class_super]
    Running test [block_scoped]
    Running test [keep_extends]
    Running test [drop_name]
    Running test [separate_name]
    Running test [static_side_effects]
    Running test [single_use]
    Running test [collapse_non_strict]
    Running test [collapse_rhs]
    Running test [collapse_rhs_static]
    Running test [property_side_effects]
    Running test [property_side_effects_static]
    Running test [unused_await]
    Running test [computed_key_side_effects]
    Running test [computed_key_generator]
    Running test [issue_4681]
    Running test [issue_4683]
    Running test [issue_4685_1]
    Running test [issue_4685_2]
    Running test [issue_4687_1]
    Running test [issue_4687_2]
    Running test [issue_4705]
```
